cpp/src/arrow/compute/expression.cc

```diff
@@ -1258,6 +1373,92 @@ struct Inequality {
+      }
+
+      const auto& lhs = Comparison::StripOrderPreservingCasts(call->arguments[0]);
+      if (!lhs.field_ref()) return expr;
+      if (*lhs.field_ref() != guarantee.target) return expr;
+
+      auto options = checked_pointer_cast<SetLookupOptions>(call->options);
+      auto unsimplified_value_set = options->value_set.make_array();
+      if (unsimplified_value_set->length() == 0) return literal(false);
+
+      Type::type value_set_type = unsimplified_value_set->type_id();
+      // Simplification for `is_in` requires that comparison kernels exist, so
+      // we skip simplification for non-primitive and non-base-binary types.
+      if (!is_integer(value_set_type) && !is_temporal(value_set_type)
+          && !is_base_binary_like(value_set_type)) {
+        return expr;
+      }
+      // For now, we abort simplification if the guarantee bound's type does not
+      // exactly match the value set's type.
+      if (guarantee.bound.type()->id() != value_set_type) return expr;
+
+      std::shared_ptr<Array>& value_set = context.is_in_value_sets[call];
+      if (!value_set) {
+        // Simplification for `is_in` requires that the value set is preprocessed.
+        // We store the prepared value set in the kernel state to avoid repeated
+        // preprocessing across calls to `SimplifyWithGuarantee`.
+        auto state = checked_pointer_cast<internal::SetLookupStateBase>(
+            call->kernel_state);
+        if (!state->sorted_and_unique_value_set) {
+          ARROW_ASSIGN_OR_RAISE(state->sorted_and_unique_value_set,
+                                PrepareIsInValueSet(unsimplified_value_set));
+        }
+        if (state->sorted_and_unique_value_set->length() == 0) {
+          context.is_in_value_sets.erase(call);
+          return literal(false);
+        }
+        value_set = state->sorted_and_unique_value_set;
+      }
+
+#define CASE(TYPE_CLASS)                                                       \
+  case TYPE_CLASS##Type::type_id:                                              \
+    result = SimplifyIsInValueSet<TYPE_CLASS##Type>(guarantee, value_set);     \
+    break;
+
+      std::variant<std::shared_ptr<Array>, bool> result;
+      switch (value_set_type) {
+        CASE(UInt8)
+        CASE(Int8)
+        CASE(UInt16)
+        CASE(Int16)
+        CASE(UInt32)
+        CASE(Int32)
+        CASE(UInt64)
+        CASE(Int64)
+        CASE(Date32)
+        CASE(Date64)
+        CASE(Time32)
+        CASE(Time64)
+        CASE(Timestamp)
+        CASE(Binary)
+        CASE(String)
+        CASE(LargeBinary)
+        CASE(LargeString)
+        default:
+          DCHECK(false);
+          return expr;
+      }
+
+#undef CASE
+
+      if (std::holds_alternative<bool>(result)) {
+        context.is_in_value_sets.erase(call);
+        return literal(std::get<bool>(result));
+      }
+      value_set = std::get<std::shared_ptr<Array>>(result);
+      return expr;
```

Comment:
These 70 lines can become a separate function -- `SimplifyIsIn` -- because we shouldn't fit all the simplifications (for all simplifiable functions) in here.